### PR TITLE
Nodal Patch Recovery Aux

### PR DESCRIPTION
**Pull request by hugary1995**
Elemental patch recovery that fits nodal values for gradient and material properties

The main work is done in NodalPatchRecovery.C which derives from AuxKernel
NodalPatchRecovery.C
  -- generates multi-index for a complete set of monomials
  -- assembles the patch stiffness matrix and load vector
  -- performs the least square fitting
  -- uses the fitted coefficients to calculate the nodal value
  -- leaves ComputeValue() abstract

Sample usage of the AuxKernel is illustrated in StressRecovery.C and NodalGradientAux.C which derives from NodalPatchRecovery and implement the ComputeValue() method.

A flag was added in AuxKernel.C to let it know whether the derived AuxKernel is using NodalPatchRecovery. If the flag is set to be true, then it won't issue error when retrieving material properties, gradients, etc.

Corresponding regression tests have been added.

refs #11880

### CODE REVIEW

**Comment by YaqiWang** [comment]
We possibly should also do a check to make sure that no patched nodal variable depends on another patched nodal variable if we do the element loop once for all aux kernels. If we keep this way, i.e. each aux kernel has its own element loop, the dependency will be fine. Sorry for my comments, this is a great capability.

File: modules/tensor_mechanics/src/auxkernels/StressRecovery.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/auxkernels/StressRecovery.C]: Some might disagree with me, but I don't really like the idea of having two separate classes for getting the stress (one for element and one for nodal values).  Could we make it so that NodalPatchRecovery can also run on element AuxVariables, and then make RankTwoScalarAux derive from it, and have it work in both cases?

File: modules/tensor_mechanics/test/tests/stress_recovery/patch/patch.i
Comment (bwspenc) [inline modules/tensor_mechanics/test/tests/stress_recovery/patch/patch.i]: Don't use superlu unless you really need it (because it's not necessarily compiled in).

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (bwspenc) [inline framework/src/auxkernels/NodalPatchRecovery.C]: I take it that this fixed the parallel problems? Are you essentially ghosting out a second layer of elements?

**Review by bwspenc — CHANGES_REQUESTED** [review]
A few things to address, but this is getting there.

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (hugary1995) [reply framework/include/auxkernels/NodalPatchRecovery.h]: I think N choose K is a quite basic and fundamental algorithm. Do you think it can go into MooseUtil?

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (hugary1995) [reply framework/src/auxkernels/NodalPatchRecovery.C]: You are right I am fixing the parallel issue here. However, this is not ghosting out a second layer.
It is a bit complicated to explain without drawing any graph. Here I am ghosting out all semilocal elements -- which is different from simply ghosting one layer out in a very subtle way.

I am also trying to get distributed mesh to work using RelationshipManager.

Comment (permcody) [reply framework/src/auxkernels/NodalPatchRecovery.C]: I recommend you stay away from semilocal elements. It's not quite what you think it is, trust me. Try searching for it in the tickets! 

The right way is with a relationship manager as you've already said. There's already existing objects that will do arbitrary numbers of layers or point neighbors and then you are good to go.

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (permcody) [inline framework/include/auxkernels/NodalPatchRecovery.h]: Wrong header - make sure you update all of them.

File: modules/tensor_mechanics/test/tests/stress_recovery/patch/patch.i
Comment (permcody) [reply modules/tensor_mechanics/test/tests/stress_recovery/patch/patch.i]: If you do, make sure you add super lu to the test spec so that we won't attempt to run this test on a machine that doesn't have it.

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (permcody) [reply framework/include/auxkernels/NodalPatchRecovery.h]: Yes, but I wouldn't want this implementation. If we add this to MOOSE it should be rewritten to be a generator instead. Building and returning all results in vectors can be _very_ space expensive.

Comment (dschwen) [reply framework/include/auxkernels/NodalPatchRecovery.h]: This is already in libMesh as `Utility::binomial(n , k)`!

Comment (dschwen) [reply framework/include/auxkernels/NodalPatchRecovery.h]: https://github.com/libMesh/libmesh/blob/master/include/utils/utility.h#L219-L237

Comment (hugary1995) [reply framework/include/auxkernels/NodalPatchRecovery.h]: @dschwen Utility::binomial(n, k) computes the number of combinations of n choose k. What I am trying to do here is to return a listing of all combinations -- maybe I should use a more descriptive name here.

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (dschwen) [reply framework/src/auxkernels/NodalPatchRecovery.C]: @hugary1995 are you trying to get _point neighbors_ here? Because that would require at least *2* additional ghosting layers with quad elements, and possibly more with tets etc! You also need to be super careful with periodic boundary conditions.

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (dschwen) [reply framework/include/auxkernels/NodalPatchRecovery.h]: Ha, right, I should have at least checked the return types...

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (permcody) [reply framework/src/auxkernels/NodalPatchRecovery.C]: This problem is solved: See how I do it for the GrainTracker:
https://github.com/idaholab/moose/blob/devel/modules/phase_field/src/postprocessors/GrainTracker.C#L55

https://github.com/idaholab/moose/blob/devel/modules/phase_field/src/relationshipmanagers/GrainTrackerHaloRM.C

Now, your use case will be a little more complex because you actually need solution data off from the ghosted element correct? For that, you'll need to create an AlgebraicRelationshipManager and we'll have to figure out exactly what we'll need to do to get solution information. This is probably also a solved problem but if not, I'm willing to help you reach that solution. Please stop by and chat when you are ready to work on this. @roystgnr will be helpful in well in telling us 1) if a ghosting functor exists that already gets solution information from n layers deep 2) Is that built into the "DefaultCoupling" ghosting functor already? Sorry, too lazy to look at the moment.

Comment (roystgnr) [reply framework/src/auxkernels/NodalPatchRecovery.C]: It depends on how you define "layers".  DefaultCoupling and PointNeighborCoupling both have set_n_levels(int) methods; if you set that to e.g. 2 then with DefaultCoupling you get side-neighbors-of-side-neighbors and with PointNeighborCoupling you get point-neighbors-of-point-neighbors.

Comment (hugary1995) [reply framework/src/auxkernels/NodalPatchRecovery.C]: @permcody I'll have time to work on this next week. What would be the best time for you to meet?

**Comment by hugary1995** [comment]
@bwspenc All touches in AuxKernel are now removed. I also derived RankTwoAux and RankTwoScalarAux from NodalPatchRecovery so they work with either nodal variables or elemental variables. The default setting in TensorMechanicsAction hasn't been changed yet -- it is still running on constant monomial.

@permcody I rolled back to manual ghosting and added TODOs in the code so that people are aware of the relationshipmanager issue. Tests are still only run on replicated mesh until we switch to relationshipmanager.

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (hugary1995) [reply framework/src/auxkernels/NodalPatchRecovery.C]: This is the only comment I did not address. Here the order means polynomial order of the least squares fitting. I think it is different from the variable order, i.e. this kernel is always run on a first lagrange variable but the polynomial order is user-defined.

**Comment by bwspenc** [comment]
A few things to address based on our discussion:

1. Change 'ploynomial_order' -> 'patch_polynomial_order', and make its default be the variable's order.
2. Add a couple of sentences in the documentation of RankTwoAux and RankTwoScalarAux to explain what happens when you have a nodal AuxVariable.
3. Test on higher order variables.

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (bwspenc) [inline framework/include/auxkernels/NodalPatchRecovery.h]: All of these methods need documentation that follows the form shown under 'Documentation' in.
http://mooseframework.org/wiki/CodeStandards/
Use the '/**' style, and describe all parameters and returns using `@param` and `@return`.

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (bwspenc) [inline framework/include/auxkernels/NodalPatchRecovery.h]: nchoosek -> nChooseK to follow MOOSE naming conventions.

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (bwspenc) [inline framework/include/auxkernels/NodalPatchRecovery.h]: Since `_order_provided` isn't used anywhere but in the constructor, I'd get rid of it.

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (bwspenc) [inline framework/src/auxkernels/NodalPatchRecovery.C]: _order(isParamValid("patch_polynomial_order") ? getParam<unsigned>("patch_polynomial_order") : (unsigned)_var.order()),

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (bwspenc) [inline framework/include/auxkernels/NodalPatchRecovery.h]: `_order` -> `_patch_polynomial_order` (to avoid confusion with variable order).

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (bwspenc) [inline framework/include/auxkernels/NodalPatchRecovery.h]: Consider getting rid of `_dim` as a member variable as well since it's also only used in the constructor.

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (bwspenc) [inline framework/src/auxkernels/NodalPatchRecovery.C]: Add a better description of this:
Polynomial order used in least squares fitting of material property over the local patch of elements connected to a given node

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (bwspenc) [inline framework/src/auxkernels/NodalPatchRecovery.C]: Would it make sense to have a check to make sure that the user didn't specify a patch polynomial order lower than the variable order?
```
   if (isParamValid("patch_polynomial_order"))
    mooseWarning("Specified 'patch_polynomial_order' is lower than the AuxVariable's order");
```

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (bwspenc) [inline framework/include/auxkernels/NodalPatchRecovery.h]: Per MOOSE code standards, change all instances of 'unsigned' -> 'unsigned int'

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (bwspenc) [inline framework/include/auxkernels/NodalPatchRecovery.h]: Either here or in the documentation for _multi_index, explain what the entries mean.

File: framework/include/auxkernels/NodalPatchRecovery.h
Comment (bwspenc) [inline framework/include/auxkernels/NodalPatchRecovery.h]: Add this to the description:
If used with elemental (monomial) AuxVariables, this class falls back on the standard techniques in AuxKernel.C to compute either constant or higher-order monomials, depending on the order of the AuxVariable.

File: modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md
Comment (bwspenc) [inline modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md]: Is the original 'volume-weighted average' statement wrong? I think it is more generally understood.

File: modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md
Comment (bwspenc) [inline modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md]: I think I would re-arrange the above text a little to convey the fact that there are really 3 main options: constant monomial, higher order monomial, and patch recovery for nodal variables. Emphasize that for the best (smoothest, most accurate) results, use the last option, but there is some additional computational cost.

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (bwspenc) [inline framework/src/auxkernels/NodalPatchRecovery.C]: A comment here indicating what we are doing with this switch would be helpful (just something simple like "Fall back on standard procedure for element variables", and "Use Zienkiewicz Zu patch recovery for nodal variables").

**Review by bwspenc — CHANGES_REQUESTED** [review]
This is looking really close.  Some nit-picky requests.

**Comment by bwspenc** [comment]
@permcody I think this is getting really close.  I'm ready to approve this as soon as he addresses my issues. I'd like to make sure we get your buy-in on this architecture before we merge this, so if you could look this over, I'd appreciate it.  We can discuss this tomorrow.

File: framework/src/auxkernels/NodalPatchRecovery.C
Comment (hugary1995) [reply framework/src/auxkernels/NodalPatchRecovery.C]: I think it is okay at least in some cases to specify a patch polynomial order higher than the variable order as we discussed earlier. I do have a check to make sure we have enough patch sample points to do the least squares fitting:

```
// before we go, check if we have enough sample points for solving the least square fitting
  if (_q_point.size() * elem_ids.size() < _multi_index.size())
    mooseError("There are not enough sample points to recover the nodal value, try reducing the "
               "polynomial order or using a higher-order quadrature scheme.");
```

File: modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md
Comment (hugary1995) [reply modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md]: I haven't heard of it before... Anyways I have rolled it back to volume-weighted average.

File: modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md
Comment (bwspenc) [inline modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md]: Try something like this:
```
* _Element Constant Monomial_ Using an AuxVariable ...
* _Elemental Higher-order Monomial_ blah blah blah...

**Review by bwspenc — CHANGES_REQUESTED** [review]

**Review by bwspenc — DISMISSED** [review]

**Review by bwspenc — APPROVED** [review]

**Pull request merged**
